golem-worker-executor-base/src/services/file_loader.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::anyhow;
+use golem_common::model::InitialComponentFileKey;
+use std::{path::PathBuf, sync::Arc};
+use tempfile::TempDir;
+use tokio::fs::File;
+use tokio::fs::{copy, create_dir_all, hard_link};
+use tokio::io::AsyncWriteExt;
+use tracing::debug;
+
+use crate::error::GolemError;
+
+use golem_service_base::service::initial_component_files::InitialComponentFilesService;
+
+/// Interface for loading files and making them available to workers.
+/// This will hardlink to a temporary directory to avoid copying files between workers. Beware
+/// that hardlinking is only possible within the same filesystem.
+// TODO: add bookkeeping for files that are no longer needed.
```

Comment:
This should be done otherwise we fill up the executor's disk eventually. 

It could be just something like a reference counting on the shared directories, decremented when a `Worker` is stopped.